core/base/iterator_factory.hpp

```diff
@@ -24,6 +24,234 @@ template <typename... Iterators>
 class zip_iterator;
 
 
+template <typename... Iterators>
+class zip_iterator_reference;
+
+
+template <typename T, typename... Ts>
+class device_tuple;
+
+
+}  // namespace detail
+}  // namespace gko
+
+
+// structured binding specializations for device_tuple, zip_iterator_reference
+namespace std {
+
+
+template <typename... Ts>
+struct tuple_size<gko::detail::device_tuple<Ts...>>
+    : integral_constant<size_t, sizeof...(Ts)> {};
+
+
+template <std::size_t I, typename... Ts>
+struct tuple_element<I, gko::detail::device_tuple<Ts...>> {
+    using type = typename tuple_element<I, tuple<Ts...>>::type;
+};
+
+
+template <typename... Iterators>
+struct tuple_size<gko::detail::zip_iterator_reference<Iterators...>>
+    : integral_constant<size_t, sizeof...(Iterators)> {};
+
+
+template <std::size_t I, typename... Iterators>
+struct tuple_element<I, gko::detail::zip_iterator_reference<Iterators...>> {
+    using type = typename iterator_traits<
+        typename tuple_element<I, tuple<Iterators...>>::type>::reference;
+};
+
+
+}  // namespace std
+
+
+namespace gko {
+
+
+/** std::get reimplementation for device_tuple. */
+template <std::size_t index, typename... Ts>
+constexpr typename std::tuple_element<index, detail::device_tuple<Ts...>>::type&
+get(detail::device_tuple<Ts...>& tuple);
+
+
+/** std::get reimplementation for const device_tuple. */
+template <std::size_t index, typename... Ts>
+constexpr const typename std::tuple_element<index,
+                                            detail::device_tuple<Ts...>>::type&
+get(const detail::device_tuple<Ts...>& tuple);
+
+
+namespace detail {
+
+
+/** simplified constexpr std::tuple reimplementation for use in device code. */
+template <typename T, typename... Ts>
+class device_tuple {
```

Comment:
I think this might have come up before, but why do we need our own implementation of `tuple`? Wouldn't providing alias for either thrust, onedpl, or std tuple work as well?